### DISCUSSION TITLE
502 Response from deployed Edge Functions deployed which behaves normally when served locally

### DISCUSSION THREAD

**Discussion by marcderhammer** [supabase/supabase#6710] [Questions] [answered]
Hoping to get some insight or help here.  

I am experiencing an issue with an Edge Function that will respond with a `502` response when deployed to the server, but it behaves normally when served locally. 

It seems to be happening when importing and using certain external modules. I understand Supabase isn't expected to debug external modules, but I was hoping to receive some help or explanations for what I'm experiencing.

To reproduce, simply create a new edge function like the following: 

```
import { serve } from "https://deno.land/std@0.131.0/http/server.ts"
import { createCanvas } from "https://deno.land/x/canvas/mod.ts"

console.log("Hello from Functions!")

serve(async (req) => {
  const { name } = await req.json()
  const data = {
    message: `Hello ${name}!`,
  }

  const canvas = createCanvas(100, 100);

  return new Response(
    JSON.stringify(data),
    { headers: { "Content-Type": "application/json" } },
  )
})
```

- Serve it locally using CLI:  `supabase functions serve <whatever>`
- Invoke it and note that it seems OK
- Deploy this same function
- Invoke it on the server and note that it returns a `502` response
```
502: Bad Gateway (DEPLOYMENT_FAILED)

The deployment failed while serving the request.
```
- The "Invocations" tab also shows this request and `502`
- Commenting out the `createCanvas` line and redeploying will make the function behave normally again.

I have experienced this same behavior with other imported libraries, this is just a simple case where it can be reproduced with a small amount of work. 

**Am I reaching some sort of memory or CPU limit that doesn't trigger locally?**  

For what it's worth (and this is likely irrelevant), my use case is generating thumbnail images for `storage`, and I was able to do so with [this](https://deno.land/x/deno_image@0.0.4) however it strips the exif data from jpegs, and I'd like to retain it, so I've been tumbling down this rabbit hole.  It'd be great to have this purely in an edge function instead of spinning up something else to do this. 

Thanks for your time!

**Comment by GaryAustin1** [comment, marked as answer]
Have you looked thru the error logs on the function in the dashboard?

You say you are invoking from "the server" so I assume you don't need cors.  If your request is using cors the first time into the function is the preflight option request.

**Reply by marcderhammer** [reply]
Thanks for your response!  Apologies for the wording, I tried invoking it via supabase-js client (in an app) and through `curl` from my pc with the same result.  

When receiving a normal response (without the "canvas" call), the `Logs` tab prints my log statements just fine, but interestingly, for the invocations which result in `502` no error logs are printed (even logs that should have printed BEFORE the "canvas" code executes.

**Reply by GaryAustin1** [reply]
I looked at the canvas repo and if I read it correctly it is 16MB.   Free script limit is 1MB and paid is 10MB.  I've not used functions enough to know if that is a direct relationship but I expect so.

**Reply by marcderhammer** [reply]
This explains it!  A memory limit.  Thank you!

**Comment by christianbm1** [comment]
Hello,

I'm also experiencing this issue. Granted, my function is much larger than @MarcDerhammer's fn. My function deploys well during local development, and it also successfully deployed when deploying through Deno's Deploy Playground service.

Is the accepted answer correct: a 502 status is being given if scripts limits are greater than 1MB (for free accounts)?

Also, is it possible to increase that limit seeing as Deno's deploy service doesn't have these restrictions?

Thanks,
Christian

**Reply by marcderhammer** [reply]
https://supabase.com/pricing

I did find this on the pricing page

screenshot as of right now: 
![image](https://user-images.githubusercontent.com/18166829/168375921-99e2f0e9-d3cc-43da-b7e3-7c6113260ad2.png)

**Comment by maximilian-lindsey** [comment]
I get the same error for a function where the entire repository is only 150 KB in size and when I bundle the function via `deno bunlde` it is only around 33 KB, so it shouldn't be the script size limit.
My guess is that I'm running into a memory limit. According to my local logs (Supabase doesn't seem to get the logs when I run it on their infra) my script needs around a 150 MB in memory and runs fine in my local Docker instance, but not on prod.
Would be great if the error messages from functions would be a bit more informative.

Does anyone here know **what the production memory limits are** for Supabase functions? 
I can't find anything about that in the docs nor on the pricing page.

**Reply by henningko** [reply]
How did you log the memory requirements of your script?

**Comment by candymandev** [comment]
Also ran into this issue today. I'm using the Deno canvas repo.

For what it's worth, I also tried deploying my function to Deno Deploy. It doesn't work there either, even after upgrading to the pro plan. I receive a "502: Bad Gateway" error (same error as Supabase) and here are the logs:

```
EvalError: Code generation from strings disallowed for this context
    at new Function (<anonymous>)
    at Rb (https://deno.land/x/canvas@v1.4.1/src/lib.js:4147:17)
    at Sb (https://deno.land/x/canvas@v1.4.1/src/lib.js:4155:15)
    at Module.<anonymous> (https://deno.land/x/canvas@v1.4.1/src/lib.js:5259:30)
    at init (https://deno.land/x/canvas@v1.4.1/src/canvaskit.ts:6:22)
    at init (https://deno.land/x/canvas@v1.4.1/src/canvas.ts:6:20)
    at https://deno.land/x/canvas@v1.4.1/mod.ts:2:22
```

There's a related issue from someone with the same error in the Deno canvas library: https://github.com/DjDeveloperr/deno-canvas/issues/21

**Reply by candymandev** [reply]
For anyone else running into this issue using canvas... I was able to use this library instead for my needs: https://deno.land/x/imagescript@1.0.11. Might be worth checking out if you don't need much like me.

**Comment by henningko** [comment]
I'm having the same issue, despite being on the Pro plan and being below the script size limit of 10MB (at 8.3MB):
```
502: Bad Gateway (ORIGIN_BOOT_RPC_ERROR)

The subhoster origin returned a failure status for the boot RPC call. Status: 500 Internal Server Error
```

I have opened the ticket with the support as nothing appears in the logs, and it works fine locally. 

I am using the following packages:
```js
import { Readability } from "https://cdn.skypack.dev/@mozilla/readability@0.4.1";
import { JSDOM } from "https://jspm.dev/npm:jsdom-deno@19.0.2";
import sanitizeHtml from "https://esm.sh/sanitize-html@2.7.0";
```
JSDOM being the culprit for the large size, but Readability won't work with `deno-dom` :-/

**Comment by slimshreydy** [comment]
How are you guys getting the size of the bundle?

**Reply by JanLoebel** [reply]
You can go into the "supabase/function/<functionName>" directory and call: `deno bundle index.ts > out.bundle` then check the out.bundle size.

**Comment by laktek** [comment]
@MarcDerhammer @henningko @candymandev Sorry for the late response on this. We just made a fix that should solve this issue.

Can you try deploying your functions and see if they work now?

**Reply by CareTiger** [reply]
Hi - i am getting the same error
![502 error on remote edge function call](https://user-images.githubusercontent.com/6777660/235379953-0efa8b6e-d7f0-4e2e-8ab1-b5ac59b20b09.png)

**Reply by laktek** [reply]
@CareTiger Can you open a support ticket from the dashboard please? https://supabase.com/dashboard/support/new

**Comment by ravi-manoj** [comment]
502: Bad Gateway (TIME_LIMIT)

The deployment failed while serving the request.


```import { serve } from "https://deno.land/std@0.168.0/http/server.ts";
import {
  SupabaseClient,
  createClient,
} from "https://esm.sh/@supabase/supabase-js@2";
import "https://deno.land/std@0.190.0/dotenv/load.ts";

async function getUser(supabaseClient: SupabaseClient, num: string) {
  const { data, error } = await supabaseClient
    .from("user")
    .select("*")
    .eq("num", num);
  if (error) throw error;
  return data;
}

serve(
  async (req) => {
    try {
      const url = new URL(req.url);
      const num = url.searchParams.get("num");
      if (num === null) throw new Error("Invalid Number");
      const supabaseClient = createClient(
        Deno.env.get("SUPABASE_URL") ?? "",
        Deno.env.get("SUPABASE_ANON_KEY") ?? "",
        {
          global: {
            headers: { Authorization: req.headers.get("Authorization")! },
          },
        }
      );
      console.log(supabaseClient);
      const data = await getUser(supabaseClient, num as string);
      return new Response(
        JSON.stringify({
          status: "200",
          result: Object.keys(data).length === 0 ? 0 : 1,
          data,
        }),
        {
          headers: { "Content-Type": "application/json" },
          status: 200,
        }
      );
    } catch (error) {
      return new Response(
        JSON.stringify({ result: 0, status: " data not found ", error }),
        {
          headers: { "Content-Type": "application/json" },
          status: 500,
        }
      );
    }
  },
  { port: 8080 }
);
```

**Comment by wax911** [comment]
Hi all, I have also been trying out supabase functions in the past month or so. At this phase mostly testing out things using postman. Request and responses are fairly light and with static data on one of the functions and all has been well till recently (a month and a half) where all the functions just return an HTTP/502 (took about 22 seconds)
```
502: Bad Gateway (TIME_LIMIT)

The deployment failed while serving the request.
```

I have some logs that tell me when the function has actually loaded and it doesn't seem to be the case (see image below): 
<img width="1272" alt="image" src="https://github.com/supabase/supabase/assets/7859175/10ff5fcb-0ebb-4339-ba59-b95ee98ed2b9">

After a few retries (randomly decides to work but responses take a lot longer) it'll respond and then starts failing again. Looking at the logs which start as soon as the function begins measuring/marking execution times for critical code paths the longest operation I can see is from the function booting up:
<img width="635" alt="image" src="https://github.com/supabase/supabase/assets/7859175/86ad01a4-e406-4586-801e-18edc4c3439c">

**Meta data**
```json
{
  "boot_time": null,
  "deployment_id": "#################_8d68d61e-941d-463d-b2aa-97344f264119_87",
  "event_type": "timeLimit",
  "execution_id": "5dece0ef-dd53-46c5-a90a-00ccfd0d4f69",
  "function_id": "8d68d61e-941d-463d-b2aa-97344f264119",
  "level": "warning",
  "project_ref": "#################",
  "region": "gcp-europe-west8",
  "timestamp": "2023-06-16T20:08:00.845661762Z",
  "version": "87"
}
```
Initially I thought something had gone wrong with the dependencies I had so I attempted to gut out everything that was starting up in any shared modules, but this has not changed the behaviour.

Unfortunately there are no error logs for me to dig deeper into the root cause of this issue beyond this high level `TIME_LIMIT`, might it be timing out on boot because of some dynamic import? I'm mainly using the following dependencies in a `_shared` module:

```json
{
  "imports": {
    "std/semver": "https://deno.land/std@0.192.0/semver/mod.ts",
    "std/uuid": "https://deno.land/std@0.192.0/uuid/mod.ts",
    "std/dotenv": "https://deno.land/std@0.192.0/dotenv/mod.ts",
    "std/testing/asserts": "https://deno.land/std@0.192.0/testing/asserts.ts",
    "std/testing/bdd": "https://deno.land/std@0.192.0/testing/bdd.ts",
    "std/testing/mock": "https://deno.land/std@0.192.0/testing/mock.ts",
    "std/testing/types": "https://deno.land/std@0.192.0/testing/types.ts",
    "x/optic": "https://deno.land/x/optic@1.3.7/mod.ts",
    "x/optic/formatters": "https://deno.land/x/optic@1.3.7/formatters/mod.ts",
    "x/optic/regex-filter": "https://deno.land/x/optic@1.3.7/filters/regExpFilter.ts",
    "x/optic/profiler": "https://deno.land/x/optic@1.3.7/logger/profileMeasure.ts",
    "x/oak": "https://deno.land/x/oak@v12.5.0/mod.ts",
    "x/upstash_redis": "https://deno.land/x/upstash_redis@v1.21.0/mod.ts",
    "esm/ua-agent-parser": "https://esm.sh/ua-parser-js@2.0.0-alpha.2/",
    "esm/supabase": "https://esm.sh/@supabase/supabase-js@2.25.0",
    "esm/growthbook": "https://esm.sh/@growthbook/growthbook@0.27.0",
    "skypack/upstash/ratelimit": "https://cdn.skypack.dev/@upstash/ratelimit@latest"
  }
}
```
